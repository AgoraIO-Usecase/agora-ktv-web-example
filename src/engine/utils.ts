import { LyricLineModel, ReactInfo, LyricToneModel, LyricModel, EnumCombo } from "./type";
import { COLORING_REACT_TIME, SECOND_PX, CURSOR_AREA_WIDTH, HIT_SCORE_THRESHOLD, FULL_MARK } from "./const";

export function getAverage(arr: number[] = []) {
  if (!arr.length) {
    return 0;
  }
  let total = arr.reduce((t, c) => t + c, 0);
  return total / arr.length;
}

export function hasPitch(lines: LyricLineModel[]) {
  for (let line of lines) {
    for (let tone of line.tones) {
      if (tone.pitch) {
        return true;
      }
    }
  }
  return false;
}

export function getTotal(lines: LyricLineModel[]) {
  return lines.length * 100;
}

export function dealPitch(lines: LyricLineModel[]) {
  let min = Number.MAX_VALUE;
  let max = Number.MIN_VALUE;

  for (let line of lines) {
    for (let tone of line.tones) {
      if (tone.pitch < min) {
        min = tone.pitch;
      }
      if (tone.pitch > max) {
        max = tone.pitch;
      }
    }
  }

  return {
    min,
    max,
  };
}

export function cutOffReact(reacts: ReactInfo[], startX: number, endX: number): ReactInfo[] {
  return reacts.map((item) => {
    let start = item.x;
    let end = item.x + item.width;
    let newStart = start;
    let newEnd = end;
    if (start < startX) {
      // 超出左侧
      newStart = startX;
    } else if (end > endX) {
      // 超出右侧
      newEnd = endX;
    }
    return {
      ...item,
      x: newStart - startX, // 重新计算x坐标
      width: newEnd - newStart,
    };
  });
}

export function filterReact(reacts: ReactInfo[], startX: number, endX: number): ReactInfo[] {
  return reacts.filter((item) => {
    const start = item.x;
    const end = start + item.width;
    if (start >= startX && end <= endX) {
      // 在屏幕中
      return true;
    } else if (start <= startX && end >= startX) {
      // 左侧部分超出屏幕
      return true;
    } else if (start <= endX && end >= endX) {
      // 右侧部分超出屏幕
      return true;
    }
    return false;
  });
}

// 判断react是否严格在tone词中
export function isReactInTone(react: ReactInfo, tone: LyricToneModel) {
  const { beginTime, duration } = tone;
  // CURSOR_AREA_WIDTH 油漆区偏移
  const toneStartX = beginTime * SECOND_PX + CURSOR_AREA_WIDTH;
  const toneEndX = toneStartX + duration * SECOND_PX;
  const reactStartX = react.x;
  const reactEndX = react.x + react.width;
  return reactStartX >= toneStartX && reactEndX <= toneEndX;
}

// 间距小于一个react的宽度;
export function isClosedReact(pre: ReactInfo, next: ReactInfo): boolean {
  const distance = COLORING_REACT_TIME * SECOND_PX;
  const preEnd = pre.x + pre.width;
  const nextStart = next.x;
  return Math.abs(nextStart - preEnd) < distance;
}

// 获取击中的react信息 (根据时间)
export function getHitInfo(lyric: LyricModel, currentTime: number): LyricToneModel {
  let tarTone = null;
  for (let line of lyric.lines) {
    for (let tone of line.tones) {
      const begin = tone.beginTime;
      const end = begin + tone.duration;
      if (begin <= currentTime && end >= currentTime) {
        tarTone = tone;
        break;
      }
    }
    if (tarTone) {
      break;
    }
  }
  return tarTone as LyricToneModel;
}

// 获取一行的平均分
export function getLineScore(line: LyricLineModel): number {
  let arr: number[] = [];
  line.tones.forEach((item) => {
    arr.push(getAverage(item.scores));
  });
  return Math.floor(getAverage(arr));
}

// 分数是否命中
export function isHitScore(score: number): boolean {
  return score >= HIT_SCORE_THRESHOLD * FULL_MARK;
}

// 获取一行的文字
export function getLineWords(line: LyricLineModel): string {
  let str = "";
  for (let tone of line.tones) {
    str += tone.word;
  }
  return str;
}

// 获取combo名称
export function getComboName(score: number): EnumCombo | undefined {
  if (score >= 60 && score < 75) {
    return EnumCombo.fair;
  } else if (score >= 75 && score < 90) {
    return EnumCombo.good;
  } else if (score >= 90) {
    return EnumCombo.excellent;
  }
  return undefined;
}
