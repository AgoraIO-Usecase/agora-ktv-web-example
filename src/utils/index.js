const { VITE_AGORA_APP_ID } = import.meta.env || {}

// string to Uint8Array
export function stringToUint8Array(str) {
  const arr = [];
  for (let i = 0, j = str.length;i < j;++i) {
    arr.push(str.charCodeAt(i));
  }
  const tmpUint8Array = new Uint8Array(arr);
  return tmpUint8Array;
}

// Uint8Array to  string
export function Uint8ArrayToString(fileData) {
  let dataString = "";
  for (let i = 0;i < fileData.length;i++) {
    dataString += String.fromCharCode(fileData[i]);
  }
  return dataString;
}

// 是否主播
export function isHost(role) {
  return role === "host";
}


export function getMusicList(role) {
  return [
    {
      songCode: "6625526605291650",
      name: "十年",
      singer: "陈奕迅",
      select: false,
    },
    {
      songCode: "6625526603528200",
      name: "七里香",
      singer: "周杰伦",
      select: false,
    },
    {
      songCode: "6654550269701680",
      name: "告白气球",
      singer: "周杰伦",
      select: false,
    },
    {
      songCode: "6625526603296890",
      name: "晴天",
      singer: "周杰伦",
      select: false,
    },
    {
      songCode: "6625526603301530",
      name: "成都",
      singer: "赵雷",
      select: false,
    },
    {
      songCode: "6625526603287770",
      name: "纸短情长",
      singer: "花粥",
      select: false,
    },
  ];
}

export const PREFIX = "/chorus/"


export const APP_INFO = [
  {
    appId: VITE_AGORA_APP_ID,
    type: "default"
  },
  {
    appId: "6bb480f77c6c458696eadb61dfc6fb76",
    type: "test"
  }
]




// 生成延迟后的AudioBuffer
export const genDelayAudioBuffer = (buffer, ms) => {
  var numChannels = buffer.numberOfChannels;
  var sampleRate = buffer.sampleRate;
  var duration = buffer.duration;

  const oldFrameCount = Math.floor(sampleRate * duration)
  const newFrameCount = Math.floor(sampleRate * (ms / 1000 + duration))

  // 创建新的audioBuffer数据
  const newAudioBuffer = new AudioContext().createBuffer(numChannels, newFrameCount, sampleRate);

  // 创建Float32的空间,作为copy数据的载体
  const anotherArray = new Float32Array(oldFrameCount);

  for (var ch = 0;ch < numChannels;ch++) {
    buffer.copyFromChannel(anotherArray, ch, 0);
    newAudioBuffer.copyToChannel(anotherArray, ch, Math.floor(ms / 1000 * sampleRate));
  }

  return newAudioBuffer
}


export async function setupSenderTransform(transceiver) {
  const isChrome =
    navigator.userAgent.indexOf("Firefox") === -1 &&
    navigator.userAgent.indexOf("Chrome") > -1;

  //@ts-ignore
  if (!transceiver || !isChrome || !window.ENABLE_ENCODED_TRANSFORM) {
    return;
  }
  const sender = transceiver.sender;

  const id = transceiver.sender.track?.id;
  if (!id) {
    return;
  }

  let streams;
  try {
    // @ts-ignore
    streams = sender.createEncodedStreams();
  } catch (e) {
    console.log(`cannot create encoded streams for track ${id}`);
    return;
  }
  const transformer = new TransformStream({
    transform: async (chunk, controller) => {
      controller.enqueue(chunk);
    },
  });

  streams.readable.pipeThrough(transformer).pipeTo(streams.writable);
}
