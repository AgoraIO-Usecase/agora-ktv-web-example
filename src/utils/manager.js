class AudioBufferManager {

  constructor(url) {
    this.url = url
    this.arrayBuffers = []
    this.audioBuffers = []
    this.audioTrack = null
    this.intervalIds = []
  }

  async deal() {
    const buffer = await this.load()
    const byteLength = buffer.byteLength
    const numParts = 10;
    const chunkSize = Math.ceil(byteLength / numParts);
    for (let i = 0;i < numParts;i++) {
      const start = i * chunkSize;
      const end = Math.min((i + 1) * chunkSize, byteLength);
      const chunk = buffer.slice(start, end);
      this.arrayBuffers.push(chunk);
    }

    console.log("this.arrayBuffers", this.arrayBuffers)
    var audioContext1 = new AudioContext();
    const audioBuffer1 = await audioContext1.decodeAudioData(this.arrayBuffers[0])

    var audioContext2 = new AudioContext();
    const audioBuffer2 = await audioContext2.decodeAudioData(this.arrayBuffers[1])

    this.audioBuffers.push(audioBuffer1)
    this.audioBuffers.push(audioBuffer2)
  }

  async createTrack() {
    const audioBuffer = this.audioBuffers.shift()
    this.audioTrack = await AgoraRTC.createBufferSourceAudioTrack({
      source: audioBuffer,
    });
    console.log("this.audioTrack", this.audioTrack)
    this.audioTrack.on("ended", () => {
      console.log("this.audioTrack end")
    })
  }

  play() {
    this.audioTrack.play();
    this.audioTrack.startProcessAudioBuffer();
    const duration = this.audioTrack.duration
    const intervalId = setInterval(() => {
      const curTime = this.audioTrack.getCurrentTime()
      if (curTime > duration - 3) {

        clearInterval(intervalId)
        return
      }
    }, 1000)
    this.intervalIds.push(intervalId)
    // getCurrentTime
  }

  async load() {
    const buffer = await fetch(this.url)
      .then((response) => response.arrayBuffer())
    // .then((arrayBuffer) => audioContext.decodeAudioData(arrayBuffer));
    return buffer;
  }


  destory(){
    this.audioTrack.stop()
    this.audioTrack.close()
    this.audioTrack = null
    this.intervalIds.forEach(id => {
      clearInterval(id)
    })
  }

}



export default AudioBufferManager
