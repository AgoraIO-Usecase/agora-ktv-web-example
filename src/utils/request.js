// http request
import axios from "axios";

let { VITE_AGORA_AUTHORIZATION, VITE_AGORA_APP_ID } = import.meta.env;
const DEFAULT_REQUEST_ID = "4135901d9c8b4f0cb09fbb51fd7508dd";
const Authorization = `Basic ${VITE_AGORA_AUTHORIZATION}`;
const songMap = new Map();
const appId = window?.appInfo?.appId || ""

// const prefix = 'http://218.205.37.50:16000'
const prefix = 'https://service-staging.agora.io/toolbox/openapi'
// const prefix = "https://api.agora.io"

let token = ""
let taskId = ""

// songCode 请求回来的MP4地址每次不一样 (做一层缓存)
// 注意MP4地址的有效期
export async function apiGetSongDetail({ requestId = DEFAULT_REQUEST_ID, songCode }) {
  try {
    if (songMap.has(songCode)) {
      return songMap.get(songCode);
    }
    const res = await axios.get(
      `https://api.agora.io/cn/v1.0/projects/${appId}/ktv-service/api/serv/song-url?requestId=${requestId}&songCode=${songCode}&lyricType=0&extTag=1`,
      { headers: { Authorization: Authorization } }
    );
    if (res.status == 200) {
      const data = res?.data?.data || {};
      songMap.set(songCode, data);
      return data;
    } else {
      return Promise.reject(res);
    }
  } catch (err) {
    return Promise.reject(err);
  }
}

export async function apiGetLyric(url = "") {
  const res = await axios({
    method: "get",
    url: url,
    responseType: "arraybuffer",
  });
  if (res.status === 200 || res.status === 0) {
    return res.data;
  } else {
    return Promise.reject(res);
  }
}

export async function apiBuildToken() {
  try {
    const res = await axios.post(
      `${prefix}/v1/projects/${appId}/rtsc/cloud-transcoder/builderTokens`,
      {
        instanceId: new Date().getTime().toString(),
        testIp: "218.205.37.50"
      },
      {
        headers: {
          "Content-Type": "application/json",
          Authorization: Authorization
        }
      },
    );
    token = res.data.tokenName
    return res.data;
  } catch (err) {
    return Promise.reject(err);
  }
}

export async function apiStartConfluence(channel) {
  try {
    const url = `${prefix}/v1/projects/${appId}/rtsc/cloud-transcoder/tasks?builderToken=${token}`
    const data = {
      services: {
        cloudTranscoder: {
          serviceType: "cloudTranscoderV2",
          config: {
            "transcoder": {
              "audioInputs": [{
                "rtc": {
                  "rtcUid": 0,
                  "rtcToken": VITE_AGORA_APP_ID,
                  "rtcChannel": channel
                }
              }], "idleTimeout": 30,
              "outputs": [{
                "audioOption": {
                  "profileType": "AUDIO_PROFILE_MUSIC_HIGH_QUALITY_STEREO",
                  "fullChannelMixer": "native-mixer-weighted"
                },
                "rtc": {
                  "rtcUid": 9527,
                  "rtcToken": VITE_AGORA_APP_ID,
                  "rtcChannel": channel + "_ad"
                },
                "dataStreamOption": {
                  "source": {
                    "dataStream": true,
                  },
                  "sink": {}
                },
                "metaDataOption": {
                  "source": {
                    "audioMetaData": true,
                  },
                  "sink": {}
                }
              }]
            }
          }
        }
      }
    }
    const res = await axios.post(url, data, {
      "Content-Type": "application/json",
      Authorization: Authorization
    })
    taskId = res.data.taskId
    return res.data
  } catch (err) {
    return Promise.reject(err);
  }
}

export async function apiStopConfluence() {
  if (!token || !taskId) {
    return
  }

  try {
    const url = `${prefix}/v1/projects/${appId}/rtsc/cloud-transcoder/tasks/${taskId}?builderToken=${token}`
    const res = await fetch(url, {
      method: "delete",
      headers: {
        "Content-Type": "application/json",
        Authorization: Authorization,
      }
    })
    token = ""
    taskId = ""
    return res.data
  } catch (err) {
    return Promise.reject(err);
  }
}
