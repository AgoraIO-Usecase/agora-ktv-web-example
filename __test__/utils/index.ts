import { apiGetLyric } from "../../src/utils/request.js";

let _dataMap = new Map();

const DEFAULT_URL = "https://accktv.sd-rtn.com/202304072151/734633a36da177df95c3a358677cb3e8/release/lyric/zip_utf8/3/251964.zip";

export const getLyricData = async (url: string = DEFAULT_URL) => {
  if (_dataMap.has(url)) {
    return _dataMap.get(url);
  }
  const res = await apiGetLyric(url);
  _dataMap.set(url, res);
  return res;
};
