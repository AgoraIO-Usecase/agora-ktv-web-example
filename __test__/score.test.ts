import { xml } from "./data/index";
import { assert, expect, test } from "vitest";
import Engine from "../src/engine/index";
import { getLyricData } from "./utils/index";

test("calc lines score", async () => {
  const res = await getLyricData();
  const engine = new Engine();
  await engine.setLyric(res);
  const lines = engine.lyric.lines.slice(0, 5);
  const lastLine = lines[lines.length - 1];
  const endTime = lastLine.beginTime + lastLine.duration;
  for (let line of lines) {
    for (let tone of line.tones) {
      let time = tone.beginTime + tone.duration / 2;
      engine.setTime(time);
      engine.setPitch(tone.pitch - 1);
    }
  }
  engine.setTime(endTime);
  let total = 0;
  let totalArr: number[] = [];
  for (let line of lines) {
    total += line.score;
    totalArr.push(line.score);
  }
  expect(lines).toMatchSnapshot();
  expect(total).toMatchInlineSnapshot("499");
  expect(totalArr).toMatchInlineSnapshot(`
    [
      99,
      100,
      100,
      100,
      100,
    ]
  `);
});
