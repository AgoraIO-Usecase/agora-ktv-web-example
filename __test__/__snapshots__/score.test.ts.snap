// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`calc lines score 1`] = `
[
  {
    "beginTime": 23.68,
    "content": "塞纳河畔左岸的咖啡",
    "duration": 2.2910000000000004,
    "score": 99,
    "tones": [
      {
        "beginTime": 23.68,
        "duration": 0.1529999999999987,
        "lang": "1",
        "pitch": 248,
        "pronounce": "",
        "scores": [
          99.65060862717586,
        ],
        "word": "塞",
      },
      {
        "beginTime": 23.833,
        "duration": 0.15300000000000225,
        "lang": "1",
        "pitch": 244,
        "pronounce": "",
        "scores": [
          99.71595831880833,
        ],
        "word": "纳",
      },
      {
        "beginTime": 23.986,
        "duration": 0.1529999999999987,
        "lang": "1",
        "pitch": 244,
        "pronounce": "",
        "scores": [
          99.78701244081006,
        ],
        "word": "河",
      },
      {
        "beginTime": 24.139,
        "duration": 0.3049999999999997,
        "lang": "1",
        "pitch": 245,
        "pronounce": "",
        "scores": [
          99.85861708807829,
        ],
        "word": "畔",
      },
      {
        "beginTime": 24.444,
        "duration": 0.15300000000000225,
        "lang": "1",
        "pitch": 252,
        "pronounce": "",
        "scores": [
          99.93128661663545,
        ],
        "word": "左",
      },
      {
        "beginTime": 24.597,
        "duration": 0.1529999999999987,
        "lang": "1",
        "pitch": 243,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "岸",
      },
      {
        "beginTime": 24.75,
        "duration": 0.20400000000000063,
        "lang": "1",
        "pitch": 242,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "的",
      },
      {
        "beginTime": 24.954,
        "duration": 0.661999999999999,
        "lang": "1",
        "pitch": 246,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "咖",
      },
      {
        "beginTime": 25.616,
        "duration": 0.3550000000000004,
        "lang": "1",
        "pitch": 234,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "啡",
      },
    ],
  },
  {
    "beginTime": 26.181,
    "content": "我手一杯品尝你的美",
    "duration": 2.6400000000000006,
    "score": 100,
    "tones": [
      {
        "beginTime": 26.181,
        "duration": 0.14699999999999847,
        "lang": "1",
        "pitch": 231,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "我",
      },
      {
        "beginTime": 26.328,
        "duration": 0.2029999999999994,
        "lang": "1",
        "pitch": 232,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "手",
      },
      {
        "beginTime": 26.531,
        "duration": 0.15300000000000225,
        "lang": "1",
        "pitch": 229,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "一",
      },
      {
        "beginTime": 26.684,
        "duration": 0.4569999999999972,
        "lang": "1",
        "pitch": 232,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "杯",
      },
      {
        "beginTime": 27.141,
        "duration": 0.15200000000000102,
        "lang": "1",
        "pitch": 205,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "品",
      },
      {
        "beginTime": 27.293,
        "duration": 0.15300000000000225,
        "lang": "1",
        "pitch": 230,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "尝",
      },
      {
        "beginTime": 27.446,
        "duration": 0.20399999999999707,
        "lang": "1",
        "pitch": 219,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "你",
      },
      {
        "beginTime": 27.65,
        "duration": 0.30600000000000094,
        "lang": "1",
        "pitch": 230,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "的",
      },
      {
        "beginTime": 28.31,
        "duration": 0.5110000000000028,
        "lang": "1",
        "pitch": 235,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "美",
      },
    ],
  },
  {
    "beginTime": 29.278,
    "content": "留下唇印的嘴",
    "duration": 2.596,
    "score": 100,
    "tones": [
      {
        "beginTime": 29.278,
        "duration": 0.30600000000000094,
        "lang": "1",
        "pitch": 240,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "留",
      },
      {
        "beginTime": 29.584,
        "duration": 0.35800000000000054,
        "lang": "1",
        "pitch": 236,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "下",
      },
      {
        "beginTime": 29.942,
        "duration": 0.3559999999999981,
        "lang": "1",
        "pitch": 299,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "唇",
      },
      {
        "beginTime": 30.298,
        "duration": 0.30400000000000205,
        "lang": "1",
        "pitch": 284,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "印",
      },
      {
        "beginTime": 30.602,
        "duration": 0.3559999999999981,
        "lang": "1",
        "pitch": 243,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "的",
      },
      {
        "beginTime": 30.958,
        "duration": 0.9160000000000004,
        "lang": "1",
        "pitch": 322,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "嘴",
      },
    ],
  },
  {
    "beginTime": 34.049,
    "content": "花店玫瑰名字写错谁",
    "duration": 2.6510000000000034,
    "score": 100,
    "tones": [
      {
        "beginTime": 34.049,
        "duration": 0.15200000000000102,
        "lang": "1",
        "pitch": 244,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "花",
      },
      {
        "beginTime": 34.201,
        "duration": 0.1529999999999987,
        "lang": "1",
        "pitch": 238,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "店",
      },
      {
        "beginTime": 34.354,
        "duration": 0.2049999999999983,
        "lang": "1",
        "pitch": 242,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "玫",
      },
      {
        "beginTime": 34.559,
        "duration": 0.35600000000000165,
        "lang": "1",
        "pitch": 243,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "瑰",
      },
      {
        "beginTime": 35.22,
        "duration": 0.15400000000000347,
        "lang": "1",
        "pitch": 233,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "名",
      },
      {
        "beginTime": 35.374,
        "duration": 0.10099999999999909,
        "lang": "1",
        "pitch": 244,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "字",
      },
      {
        "beginTime": 35.475,
        "duration": 0.1529999999999987,
        "lang": "1",
        "pitch": 234,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "写",
      },
      {
        "beginTime": 35.628,
        "duration": 0.509999999999998,
        "lang": "1",
        "pitch": 241,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "错",
      },
      {
        "beginTime": 36.138,
        "duration": 0.5620000000000047,
        "lang": "1",
        "pitch": 264,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "谁",
      },
    ],
  },
  {
    "beginTime": 36.7,
    "content": "告白气球风吹到对街",
    "duration": 2.7979999999999947,
    "score": 100,
    "tones": [
      {
        "beginTime": 36.7,
        "duration": 0.10199999999999676,
        "lang": "1",
        "pitch": 279,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "告",
      },
      {
        "beginTime": 36.802,
        "duration": 0.1529999999999987,
        "lang": "1",
        "pitch": 238,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "白",
      },
      {
        "beginTime": 36.955,
        "duration": 0.20300000000000296,
        "lang": "1",
        "pitch": 231,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "气",
      },
      {
        "beginTime": 37.158,
        "duration": 0.357999999999997,
        "lang": "1",
        "pitch": 229,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "球",
      },
      {
        "beginTime": 37.769,
        "duration": 0.1529999999999987,
        "lang": "1",
        "pitch": 207,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "风",
      },
      {
        "beginTime": 37.922,
        "duration": 0.15200000000000102,
        "lang": "1",
        "pitch": 224,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "吹",
      },
      {
        "beginTime": 38.074,
        "duration": 0.1529999999999987,
        "lang": "1",
        "pitch": 225,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "到",
      },
      {
        "beginTime": 38.227,
        "duration": 0.6090000000000018,
        "lang": "1",
        "pitch": 238,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "对",
      },
      {
        "beginTime": 38.836,
        "duration": 0.661999999999999,
        "lang": "1",
        "pitch": 235,
        "pronounce": "",
        "scores": [
          100,
        ],
        "word": "街",
      },
    ],
  },
]
`;
